File: internal/expr/unary.go
package expr

import "github.com/aland20/iad/internal"

type Unary struct {
	Operator internal.Token
	Right    Expr
}

func NewUnary(operator internal.Token, right Expr) Expr {
	return &Unary{
		Operator: operator,
		Right:    right,
	}
}

func (u *Unary) Accept(v Visitor) any {
	return v.VisitUnaryExpr(u)
}

File: src/expr/unary.rs
use crate::token::Token;

use super::{Expr, ExprVisitor, Visitor};

pub struct Unary {
    pub operator: Token,
    pub right: Box<dyn Expr>,
}

impl Unary {
    pub fn new(operator: Token, right: Box<dyn Expr>) -> Self {
        Unary { operator, right }
    }
}

impl Visitor<&dyn ExprVisitor<()>> for Unary {
    fn accept(&self, visitor: &dyn ExprVisitor<()>) {
        visitor.visit_unary_expr(self);
    }
}
